File: apps/downloadx/tests/unit/meta.test.ts
import { describe, expect, it } from 'vitest';

import {
  canResumeAgainst,
  createMeta,
  dehydrateState,
  deleteMeta,
  loadMeta,
  metaPath,
  persistMeta,
  updateMeta,
} from '../../src/meta.js';
import type { ChunkSnapshot, ProbeResult } from '../../src/types.js';
import { makeHarness } from '../helpers/config.js';

function exampleProbe(over: Partial<ProbeResult> = {}): ProbeResult {
  return {
    url: 'https://x/y.bin',
    finalUrl: 'https://x/y.bin',
    totalSize: 1024,
    acceptsRanges: true,
    etag: 'W/"v1"',
    lastModified: null,
    contentType: null,
    filename: 'y.bin',
    ...over,
  };
}

function exampleSnap(over: Partial<ChunkSnapshot> = {}): ChunkSnapshot {
  return {
    id: 'c0',
    offset: 0,
    length: 512,
    downloadedBytes: 0,
    status: 'pending',
    quality: 'good',
    retries: 0,
    ...over,
  };
}

describe('meta persistence', () => {
  it('persistMeta writes via tmp + rename (atomic) and loadMeta reads back', async () => {
    const { fs, io } = makeHarness();
    const meta = createMeta({ id: 'd1', probe: exampleProbe(), chunks: [exampleSnap()] });
    await persistMeta(io, { dir: '/dl', filename: 'y.bin' }, meta);
    expect(fs.hasFile(metaPath(io, { dir: '/dl', filename: 'y.bin' }))).toBe(true);
    // tmp file should not remain after rename.
    expect(fs.hasFile(`${metaPath(io, { dir: '/dl', filename: 'y.bin' })}.tmp`)).toBe(false);
    const loaded = await loadMeta(io, { dir: '/dl', filename: 'y.bin' });
    expect(loaded).not.toBeNull();
    expect(loaded?.id).toBe('d1');
    expect(loaded?.chunks[0]?.id).toBe('c0');
  });

  it('loadMeta returns null when the file is missing', async () => {
    const { io } = makeHarness();
    const loaded = await loadMeta(io, { dir: '/dl', filename: 'missing.bin' });
    expect(loaded).toBeNull();
  });

  it('loadMeta treats a corrupt JSON payload as missing', async () => {
    const { fs, io } = makeHarness();
    await fs.writeFile(
      metaPath(io, { dir: '/dl', filename: 'bad.bin' }),
      new TextEncoder().encode('{not json'),
    );
    const loaded = await loadMeta(io, { dir: '/dl', filename: 'bad.bin' });
    expect(loaded).toBeNull();
  });

  it('loadMeta rejects payloads with wrong schemaVersion', async () => {
    const { fs, io } = makeHarness();
    const target = metaPath(io, { dir: '/dl', filename: 'old.bin' });
    await fs.writeFile(
      target,
      new TextEncoder().encode(JSON.stringify({ schemaVersion: 2, id: 'x' })),
    );
    const loaded = await loadMeta(io, { dir: '/dl', filename: 'old.bin' });
    expect(loaded).toBeNull();
  });

  it('deleteMeta removes the file', async () => {
    const { fs, io } = makeHarness();
    const meta = createMeta({ id: 'd1', probe: exampleProbe(), chunks: [exampleSnap()] });
    await persistMeta(io, { dir: '/dl', filename: 'y.bin' }, meta);
    await deleteMeta(io, { dir: '/dl', filename: 'y.bin' });
    expect(fs.hasFile(metaPath(io, { dir: '/dl', filename: 'y.bin' }))).toBe(false);
  });

  it('updateMeta merges state/chunks and bumps updatedAt', () => {
    const meta = createMeta({
      id: 'd1',
      probe: exampleProbe(),
      chunks: [exampleSnap()],
      now: () => 1_000,
    });
    expect(meta.state).toBe('idle');
    const snaps = [exampleSnap({ status: 'completed', downloadedBytes: 512 })];
    const updated = updateMeta(meta, { state: 'completed', chunks: snaps });
    expect(updated.state).toBe('completed');
    expect(updated.chunks[0]?.status).toBe('completed');
    expect(updated.updatedAt).toBeGreaterThanOrEqual(updated.createdAt);
  });
});

describe('canResumeAgainst', () => {
  it('requires matching total size', () => {
    const meta = createMeta({ id: 'd', probe: exampleProbe({ totalSize: 1024 }), chunks: [] });
    expect(canResumeAgainst(meta, exampleProbe({ totalSize: 2048 }))).toBe(false);
  });

  it('requires matching ETag when both sides have it', () => {
    const meta = createMeta({ id: 'd', probe: exampleProbe({ etag: 'A' }), chunks: [] });
    expect(canResumeAgainst(meta, exampleProbe({ etag: 'B' }))).toBe(false);
    expect(canResumeAgainst(meta, exampleProbe({ etag: 'A' }))).toBe(true);
  });

  it('falls back to Last-Modified when ETag is absent on either side', () => {
    const meta = createMeta({
      id: 'd',
      probe: exampleProbe({ etag: null, lastModified: 'Mon, 01 Jan 2020' }),
      chunks: [],
    });
    expect(
      canResumeAgainst(meta, exampleProbe({ etag: null, lastModified: 'Mon, 01 Jan 2020' })),
    ).toBe(true);
    expect(
      canResumeAgainst(meta, exampleProbe({ etag: null, lastModified: 'Tue, 02 Jan 2020' })),
    ).toBe(false);
  });

  it('accepts when no validators exist on either side and sizes match', () => {
    const bare: ProbeResult = exampleProbe({ etag: null, lastModified: null });
    const meta = createMeta({ id: 'd', probe: bare, chunks: [] });
    expect(canResumeAgainst(meta, bare)).toBe(true);
  });
});

describe('dehydrateState', () => {
  it('rewrites downloading and probing to paused for persistence', () => {
    expect(dehydrateState('downloading')).toBe('paused');
    expect(dehydrateState('probing')).toBe('paused');
  });

  it('preserves terminal and paused states as-is', () => {
    expect(dehydrateState('paused')).toBe('paused');
    expect(dehydrateState('completed')).toBe('completed');
    expect(dehydrateState('error')).toBe('error');
    expect(dehydrateState('cancelled')).toBe('cancelled');
    expect(dehydrateState('idle')).toBe('idle');
  });
});

File: apps/downloadx_dart/test/unit/meta_test.dart
import 'package:downloadx/downloadx.dart';
import 'package:test/test.dart';

import '../helpers/mock_io.dart';

void main() {
  group('meta', () {
    test('createEmptyMeta has nulls before probe', () {
      final m = createEmptyMeta(id: 'abc', url: 'https://x/y');
      expect(m.schemaVersion, metaSchemaVersion);
      expect(m.filename, isNull);
      expect(m.totalSize, isNull);
      expect(m.state, DownloadState.idle);
      expect(m.chunks, isEmpty);
    });

    test('applyProbeToMeta fills probe-derived fields', () {
      final m = createEmptyMeta(id: 'abc', url: 'https://x/y');
      const probe = ProbeResult(
        url: 'https://x/y',
        finalUrl: 'https://x/y',
        totalSize: 1000,
        acceptsRanges: true,
        etag: 'W/"abc"',
        lastModified: null,
        contentType: 'application/octet-stream',
        filename: 'y',
        isHls: false,
      );
      applyProbeToMeta(m, probe, []);
      expect(m.filename, 'y');
      expect(m.totalSize, 1000);
      expect(m.acceptsRanges, isTrue);
      expect(m.etag, 'W/"abc"');
    });

    test('persist + load round trip', () async {
      final io = MockIo();
      final m = createMeta(
        id: 'abc',
        url: 'https://x/y',
        probe: const ProbeResult(
          url: 'https://x/y',
          finalUrl: 'https://x/y',
          totalSize: 300,
          acceptsRanges: true,
          etag: 'E1',
          lastModified: null,
          contentType: null,
          filename: 'y',
          isHls: false,
        ),
        chunks: [
          ChunkSnapshot(
            id: 'abc-c0',
            offset: 0,
            length: 300,
            downloadedBytes: 100,
            status: ChunkStatus.downloading,
            quality: ChunkQuality.good,
            retries: 1,
          ),
        ],
      );
      final loc = MetaLocator(dir: '/cache', id: 'abc');
      await persistMeta(io, loc, m);
      final loaded = await loadMeta(io, loc);
      expect(loaded, isNotNull);
      expect(loaded!.totalSize, 300);
      expect(loaded.chunks.length, 1);
      expect(loaded.chunks.first.downloadedBytes, 100);
      // downloading state is dehydrated to paused only when the engine persists;
      // here we persisted as-is, so it remains downloading.
      expect(loaded.chunks.first.status, ChunkStatus.downloading);
    });

    test('listMetaFiles finds only matching sidecars', () async {
      final io = MockIo();
      final loc = MetaLocator(dir: '/cache', id: 'abc');
      await persistMeta(io, loc, createEmptyMeta(id: 'abc', url: 'u'));
      await io.writeFile('/cache/not-a-meta.txt', [1, 2, 3]);
      final metas = await listMetaFiles(io, '/cache');
      expect(metas.length, 1);
      expect(metas.first.id, 'abc');
    });

    test('canResumeAgainst: etag match', () {
      final m = createEmptyMeta(id: 'a', url: 'u')
        ..totalSize = 100
        ..etag = 'E1';
      const probe = ProbeResult(
        url: 'u',
        finalUrl: 'u',
        totalSize: 100,
        acceptsRanges: true,
        etag: 'E1',
        lastModified: null,
        contentType: null,
        filename: 'f',
        isHls: false,
      );
      expect(canResumeAgainst(m, probe), isTrue);
    });

    test('canResumeAgainst: etag mismatch blocks resume', () {
      final m = createEmptyMeta(id: 'a', url: 'u')
        ..totalSize = 100
        ..etag = 'E1';
      const probe = ProbeResult(
        url: 'u',
        finalUrl: 'u',
        totalSize: 100,
        acceptsRanges: true,
        etag: 'E2',
        lastModified: null,
        contentType: null,
        filename: 'f',
        isHls: false,
      );
      expect(canResumeAgainst(m, probe), isFalse);
    });

    test('canResumeAgainst: size mismatch blocks resume', () {
      final m = createEmptyMeta(id: 'a', url: 'u')..totalSize = 100;
      const probe = ProbeResult(
        url: 'u',
        finalUrl: 'u',
        totalSize: 200,
        acceptsRanges: true,
        etag: null,
        lastModified: null,
        contentType: null,
        filename: 'f',
        isHls: false,
      );
      expect(canResumeAgainst(m, probe), isFalse);
    });

    test('dehydrateState maps downloading/probing to paused', () {
      expect(dehydrateState(DownloadState.downloading), DownloadState.paused);
      expect(dehydrateState(DownloadState.probing), DownloadState.paused);
      expect(dehydrateState(DownloadState.completed), DownloadState.completed);
      expect(dehydrateState(DownloadState.error), DownloadState.error);
    });

    test('corrupt meta loads as null', () async {
      final io = MockIo();
      await io.writeFile('/cache/bad$metaExt', 'not json'.codeUnits);
      final loaded = await loadMeta(io, MetaLocator(dir: '/cache', id: 'bad'));
      expect(loaded, isNull);
    });
  });
}
